### ISSUE TITLE
Panel widgets should expose an Accessible object to at-spi

### ISSUE THREAD

**Issue by sthibaul** [mate-desktop/mate-panel#835]
#### Steps to reproduce the behaviour

- Run a mate desktop
- Run orca, enable mouse review ("Speak object under mouse" option in its general configuration panel)
- Move mouse to mate panel wigets, for instance here the clock

#### Expected behaviour

- Orca should speak the panel widget content, in the instance here the clock date & time

#### Actual behaviour

- Orca just says "panel"

#### MATE general version

1.20

#### Package version

1.20.1

#### Linux Distribution

Debian buster

#### Additional information

The issue is that mate-panel widgets do not expose Accessible objects through at-spi. When looking in accerciser, we see mate-panel expose the panel structure (thus why Orca says "panel"), but the final widgets are not exposed.

Of course, for proper complete support, probably each and every panel widget needs to be fixed to expose its content through at-spi. Perhaps mate-panel should at least put the name of the widget in the panel description (just like it does with `panel_a11y_set_atk_name_desc` for "Top Panel"), so at least the user knows which widget it is. We can then work on adding support to widgets, at least the most commonly used.

**Comment by sthibaul** [comment]
Launchers notably do have an accessible description that Orca can read.

**Comment by sthibaul** [comment]
Here is a very crude attempt [patch.txt](https://github.com/mate-desktop/mate-panel/files/2197495/patch.txt) to do this which does work:

- what needs to wear the atk name is the socket since that's the innermost object that Orca sees
- what knows the name of the applet is apparently the MatePanelAppletFrame. For now I just took the id, perhaps there is a better name to be taken, ideally it would be translated.

that's why I had to make `mate_panel_applet_frame_dbus_set_id_descr` get the name from a call to `mate_panel_applet_frame_get_id`, and then call `mate_panel_applet_container_set_id` which calls `panel_a11y_set_atk_name_desc` on the socket.

Also, I had to call mate_panel_applet_frame_dbus_set_id_descr from `mate_panel_applet_frame_dbus_init_properties`, which is late enough for applet_info to be available.

Eventually, I copied over panel_a11y_* just to get it linking, otherwise linking mate-panel-test-applets was failing. I guess we could put these functions to a better place.

How do you feel this could be improved to be integrated?

**Comment by alexarnaud** [comment]
@lukefromdc @raveit65  What are your opinions on this?

@muktupavels It could maybe benefit also to GNOME Panel.

Best regards,
Alex.

**Comment by lukefromdc** [comment]
Would the panel applets also need work to make this function for everything in the panel? Also look at whether this behaves differently for in-process and out-of-process applets. Certainly this should be fixed.

**Comment by muktupavels** [comment]
I have no idea how this should work... Is it enough to set applet name? If so panel should be able to automatically set name and/or description that you can see in Add to Panel dialog. If it should speak content, for example, telling current temperature then this should be fixed in each applet...

This probably even does not need extra API.

`panel_a11y_get_is_a11y_enabled` function does not make any sense to me (yes I know, it exists also in gnome-panel). It gets accessible object and checks if it is really accessible object... So if it returns NULL it is false, otherwise true, because I am pretty sure that gtk does not return invalid / non AtkObject objects...

```
+	if (!panel_a11y_get_is_a11y_enabled (widget))
+		return;
+
+	aobj = gtk_widget_get_accessible (widget);
+	if (!aobj)
+		return;
```

Basically both ifs do same.

In any case I don't see how this can benefit `gnome-panel` unless someone wants open merge requests with needed changes.

**Comment by sthibaul** [comment]
> Would the panel applets also need work to make this function for everything in the panel?

Well, yes, to get all applet details, the applets themselves need to implement accessibility information.
The patch I have posted is to at least make the panel provide the applet name, so that in case an applet does not implement accessibility, at least we have its name to know what it is.

> whether this behaves differently for in-process and out-of-process applets

For the purpose here, it's not actually different: in both cases the applet provides accessibility information, and the screen reader will find it, whatever the process provides it.

> Is it enough to set applet name?

It's not enough for a complete implementation, but it is a beginning: at least we get to know what applets are there.

> If it should speak content, for example, telling current temperature then this should be fixed in each applet...

Yes.

Ok for panel_a11y_get_is_a11y_enabled.

For panel_a11y_set_atk_name_desc, where should we put it to avoid duplicating it in both gnome-panel and panel-applet-container?

**Comment by muktupavels** [comment]
Well sounds like this should be fixed in applets. How can panel know that applet has not set atk name and/or description? Especially if you can do it on any widget?

> For panel_a11y_set_atk_name_desc, where should we put it to avoid duplicating it in both gnome-panel and panel-applet-container?

In your patch just move function content to `mate_panel_applet_container_set_id`? But I would not spend time trying to get it work automatically, simply start with most important applets and do right thing from start. :)

**Comment by sthibaul** [comment]
> How can panel know that applet has not set atk name and/or description?

It does not need to know, it can set something, and let the applet overwrite it.

> In your patch just move function content to mate_panel_applet_container_set_id

Ok

> I would not spend time trying to get it work automatically, simply start with most important applets and do right thing from start.

But it means that for third party applets (which probably won't have accessibility information) we won't have any information.

**Comment by muktupavels** [comment]
On Thu, Jul 26, 2018 at 4:43 PM, Samuel Thibault <notifications@github.com>
wrote:

> How can panel know that applet has not set atk name and/or description?
>
> It does not need to know, it can set something, and let the applet
> overwrite it.
>
I think in this case panel could overwrite what applet has set, depending
where and when it would happen...

> In your patch just move function content to mate_panel_applet_container_
> set_id
>
> Ok
>
> I would not spend time trying to get it work automatically, simply start
> with most important applets and do right thing from start.
>
> But it means that for third party applets (which probably won't have
> accessibility information) we won't have any information.
>
Yes, but currently it does not work at all, right? I don't see that as big
problem and those 3rd party applet also should be fixed.

-- 
Alberts Muktupāvels

**Comment by alexarnaud** [comment]
@cwendling What do you think of this ?

I agree with the Samuel point of view. Having a fallback on each widget is a minimum to have because user of feature like mouse review that have third-party applet would like to have the applet name read by Orca.
Having nothing read could let user imagine that there are on the previous spoken aplet and so click on the wrong applet.

Best regards.

**Comment by cwendling** [comment]
I think that we should implement @sthibaul's suggestion, so that we have _something_ to present.
Yes, this is a fallback in case there is no better information, but for the moment there *is* *no* information for any out-of-process applets, and there might not even be if we implement what's needed to expose out-of-process a11y if that applet doesn't provide it.

So yes, this is not the ultimate and perfect solution, but it's similar to having a placeholder icon if you don't have anything better: if you only display an icon and don't show a placeholder if it's missing, you just end up showing nothing, which well, isn't really visible.  I'm just trying to suggest an analogy here, that it'll always be better to present *something*, even if it's just a fallback in case something else went wrong.

Ultimately, what should happen is:
* a kind of "bridge" is implemented so mate-panel can point the AT technologies to the applet's accessible tree as children of the panel (this requires communicating the information between the panel and the applet, and adding a way of telling ATK the child is in a foreign process);
* each applet implements proper accessible tree and exposes it so the panel can do the above.

This would be great, and discussing with @sthibaul suggests it might not be as challenging as it may seem, but it's definitely not trivial.
And it won't fix the case where an applet is not exposing the proper information at all, in which case we *still* should provide a sensible default so the user can tell that there is something and *what* it is, even if nothing better.

**Comment by lukefromdc** [comment]
Would porting applets to in-process (as GNOME did for gnome-applets) fix this? That
needs to be done anyway for Wayland.

**Comment by cwendling** [comment]
> Would porting applets to in-process (as GNOME did for gnome-applets) fix this?

It should fix the first part, yes.  Then all what should be needed is for the various applets to report sensible a11y info if it's not already the case.

Though, wasn't separation designed for safety/whatnot?  and is there really no solution for surface sharing or proxying in Wayland?

Anyway, yes, if we were using in-process applets it'd simplify the situation quite a bit.

**Comment by sthibaul** [comment]
in-process applets would still leave no information about the applet name if the applet does not provide any accessibility support.

**Comment by cwendling** [comment]
Yes, as I tried to say earlier, the panel providing a name is a good idea in all cases, as it'd be used if nothing better is reported, and we can't rely 100% on all applets being well-behaved.
So no matter how we improve everything else, this should still land.

**Comment by lukefromdc** [comment]
As of now the clock, fish (no doubt designed as an example for such things), wncklet (window list etc), and tray can all be build in-process or out.

**Comment by johndescs** [comment]
I've reworked @sthibaul's patch, works well. See attachment. Any comment? :)
[0001-Provide-a-fallback-accessibility-string-for-applets.txt](https://github.com/mate-desktop/mate-panel/files/3879134/0001-Provide-a-fallback-accessibility-string-for-applets.txt)

**Comment by alexarnaud** [comment]
@johndescs Could you please open a merge request for this one?

**Comment by raveit65** [comment]
@cwendling 
Can this be closed?

**Comment by sthibaul** [comment]
It seems to me that the current state (tested here with MATE 1.26) is nice, e.g. the volume, network, and time applets do get some feedback with Orca
